cumulusci/util/xml/metadata_tree.py

```diff
@@ -0,0 +1,222 @@
+
+        If the parent element (self) already one or more children that match,
+        the new element follows the last one.
+
+        Otherwise, the new element goes to the bottom of the parent (self).
+
+        >>> Package = metadata_tree.fromstring(b"""<?xml version="1.0" encoding="UTF-8"?>
+        ... <Package xmlns="http://soap.sforce.com/2006/04/metadata">
+        ...     <types>
+        ...         <members>CaseReason</members>
+        ...         <members>CaseStatus</members>
+        ...         <members>CaseType</members>
+        ...         <name>StandardValueSet</name>
+        ...     </types>
+        ...     <version>46.0</version>
+        ... </Package>""")
+        >>> Package.types.append("members", "CaseOfBeer")
+        <members>CaseOfBeer</members> element
+        >>> print(Package.types.tostring())
+        <types xmlns="http://soap.sforce.com/2006/04/metadata">
+            <members>CaseReason</members>
+            <members>CaseStatus</members>
+            <members>CaseType</members>
+            <members>CaseOfBeer</members>
+            <name>StandardValueSet</name>
+        </types>
+        '''
+        newchild = self._create_child(tag, text)
+        same_elements = self._element.findall(self._add_namespace(tag))
+        if same_elements:
+            last = same_elements[-1]
+            index = self._element.index(last)
+            self._element.insert(index + 1, newchild._element)
+        else:
+            self._element.append(newchild._element)
+        return newchild
+
+    def insert(self, index: int, tag: str, text: str = None):
+        """Insert at a particular index.
+
+        Tag and text can be supplied. Return value is the new element."""
+        newchild = self._create_child(tag, text)
+        self._element.insert(index, newchild._element)
+        return newchild
+
+    def insertBefore(self, oldElement: "MetadataElement", tag: str, text: str = None):
+        """Insert before some other element
+
+        Tag and text can be supplied. Return value is the new element."""
+        index = self._element.index(oldElement._element)
+        return self.insert(index, tag, text)
+
+    def insertAfter(self, oldElement: "MetadataElement", tag: str, text: str = None):
+        """Insert after some other element
+
+        Tag and text can be supplied. Return value is the new element."""
+
+        index = self._element.index(oldElement._element)
+        return self.insert(index + 1, tag, text)
+
+    def remove(self, metadata_element: "MetadataElement") -> None:
+        """Remove an element from its parent (self)"""
+        self._element.remove(metadata_element._element)
+
+    def find(self, tag, text=None):
+        """Find a single direct child-elements with name `tag`"""
+        return next(self._findall(tag, text), None)
+
+    def findall(self, tag, text=None):
+        """Find all direct child-elements with name `tag`"""
+        return list(self._findall(tag, text))
```

Comment:
It's a pity to lose the ability to do more complex XPath queries -- but I don't know how we would add the namespaces without parsing xpath ourselves

Reply:
Now that these things are objects it crossed my mind to use something like ObjectPath or JSONPath but so far list expressions are working reasonably well.

Reply:
I think with the new matching feature we cover 90% of the use-cases. Metadata files are not recursive and are not super-deeply nested either.